common/logger/src/tests/struct_log.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    prelude::*, set_struct_logger, struct_log::InitLoggerError, LoggingField, StructLogSink,
+    StructuredLogEntry,
+};
+use chrono::{DateTime, Utc};
+use serde_json::{Map, Value};
+use std::{
+    io,
+    sync::mpsc::{self, Receiver, SyncSender},
+};
+
+struct StreamStructLog {
+    sender: SyncSender<StructuredLogEntry>,
+}
+
+impl StreamStructLog {
+    pub fn start_new() -> io::Result<(Self, Receiver<StructuredLogEntry>)> {
+        let (sender, receiver) = mpsc::sync_channel(1024);
+        Ok((Self { sender }, receiver))
+    }
+}
+
+impl StructLogSink for StreamStructLog {
+    fn send(&self, entry: StructuredLogEntry) {
+        if let Err(e) = self.sender.send(entry) {
+            // Use log crate macro to avoid generation of structured log in this case
+            // Otherwise we will have infinite loop
+            log::error!("Failed to send structured log: {}", e);
+        }
+    }
+}
+
+fn set_test_struct_logger() -> Result<Receiver<StructuredLogEntry>, InitLoggerError> {
+    let (logger, receiver) = StreamStructLog::start_new().map_err(InitLoggerError::IoError)?;
+    let logger = Box::leak(Box::new(logger));
+    set_struct_logger(logger).map(|_| receiver)
+}
+
+fn as_object(value: Value) -> Map<String, Value> {
+    value.as_object().unwrap().clone()
+}
+
+fn val(value: &Map<String, Value>, key: &str) -> Value {
```

Comment:
you probably want to have a `impl Map<String, Value>` as helper functions

Reply:
Can you `impl Map<String, Value>` outside of the Map's crate?

Reply:
nvm, just made a trait for it

Reply:
I think you can? I only remember you can't impl for a trait outside the crate

Reply:
I had to create a trait for it, as you can't impl for the struct outside of its crate.